gateway/ln-gateway/src/events.rs

```diff
@@ -22,4 +32,288 @@ pub const ALL_GATEWAY_EVENTS: [EventKind; 11] = [
+            // included. Return all events.
+            if start_index > 0 {
+                return all_events;
+            }
+        }
+
+        end_position = start_position;
+        start_position = start_position.saturating_sub(BATCH_SIZE);
+    }
+
+    let remaining_log = end_position.distance(start_position);
+    if let Some(batch_size) = remaining_log {
+        let batch = client.get_event_log(Some(start_position), batch_size).await;
+        let (start_index, end_index) = get_index_bounds(&batch, start, end);
+
+        if end_index > 0 {
+            let events = &batch[start_index..end_index];
+            all_events.append(&mut events.to_vec());
+        }
+    }
+
+    all_events
+}
+
+/// Binary searches the `PersistedLogEntry` slice for the start timestamp and
+/// end timestamp and returns the index of each. If an exact match is not found,
+/// the index where start or end could be inserted to maintain ordering
+/// is returned.
+fn get_index_bounds(
+    batch: &[PersistedLogEntry],
+    start: SystemTime,
+    end: SystemTime,
+) -> (usize, usize) {
+    let timestamps = batch.iter().map(|e| e.timestamp).collect::<Vec<_>>();
+    let start_micros = start
+        .duration_since(UNIX_EPOCH)
+        .expect("before unix epoch")
+        .as_micros() as u64;
+    let index = timestamps.binary_search(&start_micros);
+    let start_index = match index {
+        Err(index) | Ok(index) => index,
+    };
+
+    let end_micros = end
+        .duration_since(UNIX_EPOCH)
+        .expect("before unix epoch")
+        .as_micros() as u64;
+    let index = timestamps.binary_search(&end_micros);
+    let end_index = match index {
+        Err(index) | Ok(index) => index,
+    };
+
+    (start_index, end_index)
+}
+
+/// Filters the given `PersistedLogEntry` slice by the `EventKind` and
+/// `ModuleKind`.
+pub(crate) fn filter_events(
+    all_events: &[PersistedLogEntry],
+    event_kind: EventKind,
+    module_kind: ModuleKind,
+) -> Vec<&PersistedLogEntry> {
+    let events = all_events
+        .iter()
+        .filter(|e| {
+            if let Some((m, _)) = &e.module {
+                e.event_kind == event_kind && *m == module_kind
+            } else {
+                false
+            }
+        })
+        .collect::<Vec<_>>();
+    drop(event_kind);
+    drop(module_kind);
+    events
+}
+
+/// This function computes a "nested loop join" by first computing the cross
+/// product of the start event vector and the success/failure event vectors. The
+/// resulting cartesian product is then filtered according to the join predicate
+/// supplied in the parameters.
+///
+/// This function is intended for small data sets. If the data set relations
+/// grow, this function should implement a different join algorithm or be moved
+/// out of the gateway.
+pub(crate) fn join_events<Start, Success, Failure>(
+    start_events: &Vec<&PersistedLogEntry>,
```

Comment:
No function should ever need to take `&Vec<_>`. `&[_]` is always better, as `&vec` works for it, along with `&slice`, while it's not like it the other way around. Kind of like `&str` is always better arg type than `&String`.